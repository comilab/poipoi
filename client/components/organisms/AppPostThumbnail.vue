<template lang="pug">
figure.relative
  AppSquare.w-full
    img.object-cover.border.rounded.select-none.pointer-events-none(
      v-if="showThumbnail",
      :src="post.images[0].paths.small"
    )
    .relative.border.rounded.p-4(
      v-else
    )
      figcaption.figcaption.text-sm.select-none.overflow-hidden(
        :class="{ 'has-title': post.title }"
      )
        .text-base.font-bold.truncate {{ post.title }}
        .caption {{ post.caption }}
  .absolute.top-0.px-1.py-2
    AppPostBadges(
      :post="post",
      icon-only
    )
  .absolute.right-0.bottom-0.text-right
    AppButton.border.mr-2.mb-2(
      v-if="post.imagesCount && post.showImagesList",
      type="badge",
      color="white",
      size="sm",
      :icon="faImages"
    ) {{ post.imagesCount }}
    AppButton.border.mr-2.mb-2(
      v-if="post.textCount",
      type="badge",
      color="white",
      size="sm"
    ) {{ formattedTextCount }} å­—
  .absolute.inset-0.bg-white.bg-opacity-75.opacity-0.hover_opacity-100.transition-opacity.duration-300.border.rounded.p-4
    figcaption.figcaption.text-sm.select-none.overflow-hidden(
      :class="{ 'has-title': post.title }"
    )
      .text-base.font-bold.truncate {{ post.title }}
      .caption {{ post.caption }}
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'
import { faImages } from '@fortawesome/free-solid-svg-icons'

import Post from '~/models/Post'
import AppSquare from '~/components/atoms/AppSquare.vue'
import AppButton from '~/components/atoms/AppButton.vue'
import AppPostBadges from '~/components/organisms/AppPostBadges.vue'

export default defineComponent({
  components: {
    AppSquare,
    AppButton,
    AppPostBadges
  },
  props: {
    post: {
      type: Post,
      required: true
    }
  },
  setup (props) {
    const showThumbnail = computed(() => {
      return props.post.imagesCount && props.post.showThumbnail
    })

    const formattedTextCount = computed(() => {
      return new Intl.NumberFormat().format(props.post.textCount)
    })

    return {
      showThumbnail,
      formattedTextCount,
      faImages
    }
  }
})
</script>

<style lang="postcss" scoped>
.figcaption {
  & .caption {
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 4;

    @screen md {
      -webkit-line-clamp: 6;
    }

    @screen lg {
      -webkit-line-clamp: 5;
    }

    @screen xl {
      -webkit-line-clamp: 7;
    }
  }

  &.has-title .caption {
    -webkit-line-clamp: 3;

    @screen md {
      -webkit-line-clamp: 5;
    }

    @screen lg {
      -webkit-line-clamp: 4;
    }

    @screen xl {
      -webkit-line-clamp: 6;
    }
  }
}
</style>
