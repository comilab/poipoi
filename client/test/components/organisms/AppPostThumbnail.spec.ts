import { render } from '@testing-library/vue'

import AppPostThumbnail from '~/components/organisms/AppPostThumbnail.vue'
import imageFactory from '~/test/factories/image'
import postFactory from '~/test/factories/post'

describe('components/organisms/AppPostThumbnail', () => {
  let options: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true
      }
    }
  })

  test('renders correctly', () => {
    options.props.post = postFactory.build({
      pinned: true,
      rating: 'nsfw',
      scope: 'public',
      publishStart: null,
      publishEnd: null,
      title: 'title',
      caption: 'caption',
      text: 'foo',
      showThumbnail: true,
      showImagesList: true,
      images: imageFactory.buildList(1, {
        publicPaths: {
          original: 'original',
          large: 'large',
          medium: 'medium',
          small: 'small'
        }
      })
    })
    const { container } = render(AppPostThumbnail, options)

    expect(container).toMatchSnapshot()
  })

  describe('thumbnail', () => {
    test('画像があってshowThumbnailがtrueなら表示', () => {
      options.props.post = postFactory.build({
        images: imageFactory.buildList(1),
        showThumbnail: true
      })
      const { queryByRole } = render(AppPostThumbnail, options)

      expect(queryByRole('img')).toBeInTheDocument()
    })

    test('画像があってもshowThumbnailがfalseなら非表示', () => {
      options.props.post = postFactory.build({
        images: imageFactory.buildList(1),
        showThumbnail: false
      })
      const { queryByRole } = render(AppPostThumbnail, options)

      expect(queryByRole('img')).not.toBeInTheDocument()
    })

    test('画像がなければ非表示', () => {
      options.props.post = postFactory.build({
        images: [],
        showThumbnail: true
      })
      const { queryByRole } = render(AppPostThumbnail, options)

      expect(queryByRole('img')).not.toBeInTheDocument()
    })
  })
})
