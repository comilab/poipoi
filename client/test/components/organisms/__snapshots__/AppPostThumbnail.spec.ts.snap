// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppPostThumbnail renders correctly 1`] = `
<div>
  <figure
    class="relative"
  >
    <div
      class="is-square relative w-full"
    >
      <div
        class="content absolute inset-0"
      >
        <img
          class="object-cover border rounded select-none pointer-events-none"
          src="undefinedsmall"
        />
      </div>
    </div>
    <div
      class="absolute top-0 px-1 py-2"
    >
      <ul
        class="flex"
      >
        <li
          class="mx-1"
        >
          <div
            aria-label="固定"
            class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none pointer-events-none bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm rounded-full"
            data-testid="button"
          >
            <div>
              <fontawesomeicon-stub
                class="fa-fw"
                data-testid="icon"
                icon="[object Object]"
              />
            </div>
            <div
              class=""
              data-testid="text"
            />
          </div>
        </li>
        <li
          class="mx-1"
        >
          <div
            class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none pointer-events-none bg-red-700 text-white hover_bg-red-800 text-sm px-2 py-1 rounded-full"
            data-testid="button"
          >
            <div>
              <fontawesomeicon-stub
                class="fa-fw mr-2"
                data-testid="icon"
                icon="[object Object]"
              />
            </div>
            <div
              class=""
              data-testid="text"
            >
              <span
                style="display: none;"
              >
                閲覧注意
              </span>
            </div>
          </div>
        </li>
        <!---->
        <!---->
      </ul>
    </div>
    <div
      class="absolute right-0 bottom-0 text-right"
    >
      <div
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none border mr-2 mb-2 pointer-events-none bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 text-sm px-2 py-1 rounded"
        data-testid="button"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw mr-2"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class=""
          data-testid="text"
        >
          1
        </div>
      </div>
      <div
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none border mr-2 mb-2 pointer-events-none bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 text-sm px-2 py-1 rounded"
        data-testid="button"
      >
        <!---->
        <div
          class=""
          data-testid="text"
        >
          3 字
        </div>
      </div>
    </div>
    <div
      class="absolute inset-0 bg-white bg-opacity-75 opacity-0 hover_opacity-100 transition-opacity duration-300 border rounded p-4"
    >
      <figcaption
        class="figcaption text-sm select-none overflow-hidden has-title"
      >
        <div
          class="text-base font-bold truncate"
        >
          title
        </div>
        <div
          class="caption"
        >
          caption
        </div>
      </figcaption>
    </div>
  </figure>
</div>
`;
